src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,494 @@
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code ikm} is null
+         */
+        public Builder addIKM(SecretKey ikm) {
+            Objects.requireNonNull(ikm, "ikm must not be null");
+            ikms.add(ikm);
+            return this;
+        }
+
+        /**
+         * Adds input keying material to the builder.
+         * <p>
+         * {@code addIKM} may be called when the input keying material value is to
+         * be assembled piece-meal or if part of the IKM is to be supplied by a
+         * hardware crypto device. This method appends to the existing list of
+         * values or creates a new list if there is none yet.
+         *
+         * @param ikm
+         *     the input keying material value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code ikm} is null
+         */
+        public Builder addIKM(byte[] ikm) {
+            Objects.requireNonNull(ikm, "ikm must not be null");
+            if (ikm.length != 0) {
+                return addIKM(new SecretKeySpec(ikm, "Generic"));
+            } else {
+                return this;
+            }
+        }
+
+        /**
+         * Adds a salt to the builder.
+         * <p>
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there is none yet.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(SecretKey salt) {
+            Objects.requireNonNull(salt, "salt must not be null");
+            salts.add(salt);
+            return this;
+        }
+
+        /**
+         * Adds a salt to the builder.
+         * <p>
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there is none yet.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(byte[] salt) {
+            Objects.requireNonNull(salt, "salt must not be null");
+            if (salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "Generic"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Returns a builder for building {@code Extract} and
```

Comment:
Modify `builder` to `{@code Builder} object`.

Reply:
You can also append the `object` word here.

Reply:
I meant "returns a Builder object". That said, `Builder` is already a noun, much more like a name than `Extract`. So you can ignore my suggestion. Sorry.